rest/oidc_api_test.go

```diff
@@ -1,139 +1,612 @@
 package rest
 
 import (
-	"io"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
 	"net/url"
-	"strings"
+	"sync"
 	"testing"
+	"time"
 
 	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/gorilla/mux"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2/jwt"
 )
 
-func TestGetOIDCCallbackURL(t *testing.T) {
-	var (
-		wellKnownPath           = "/.well-known/openid-configuration"
-		headerLocation          = "Location"
-		authProvider            = "provider"
-		authRedirectURI         = "redirect_uri"
-		providerNameGoogle      = "google"
-		providerNameSalesforce  = "salesforce"
-		clientIDGoogle          = "client.id.google"
-		clientIDSalesforce      = "client.id.salesforce"
-		validationKeySalesforce = "validation.key.salesforce"
-		validationKeyGoogle     = "validation.key.google"
-	)
-
-	// Mock OpenID Connect Provider Discovery Endpoints
-	var issuerGoogle, issuerSalesforce string
-	mux := http.NewServeMux()
-	mux.HandleFunc("/google"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://accounts.google.com/o/oauth2/v2/auth", 
-        	"token_endpoint": "https://oauth2.googleapis.com/token",
-			"jwks_uri": "https://www.googleapis.com/oauth2/v3/certs", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerGoogle))
-		require.NoError(t, err, "Failed to write response body for google provider")
-	})
-	mux.HandleFunc("/salesforce"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://login.salesforce.com/services/oauth2/authorize",
-        	"token_endpoint": "https://login.salesforce.com/services/oauth2/token",
-			"jwks_uri": "https://login.salesforce.com/id/keys", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerSalesforce))
-		require.NoError(t, err, "Failed to write response body for salesforce provider")
+// mockOptions keeps options required for mocking
+// auth code flow responses.
+type mockOptions map[string]interface{}
+
+var (
+	optionsOnce sync.Once
+	options     mockOptions
+)
```

Comment:
As was commented earlier, we should avoid global variables unless absolutely needed.  It looks like you've just moved the previous globals (issuerGoogle, issuerFacebook) into this struct.  

Ben already outlined a recommended approach on the other PR (https://github.com/couchbase/sync_gateway/pull/4558/files#r420686763) - we should be doing that here.

Reply:
Makes much sense, thank you for clarifying this. I've refactored the mockAuthServer code quite a bit to avoid these global scoped variable.